### PR TITLE
Add std::log1p for c10::complex

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Note that `std::log1p` is not a thing for `std::complex`, so we have to manually implement it.

### CODE REVIEW

File: c10/util/complex_math.h [hunk 1]
```diff
@@ -212,6 +212,28 @@ C10_HOST_DEVICE inline c10::complex<T> atanh(const c10::complex<T> &x) {
 #endif
 }
 
+// We also support some math functions that is not supported by standard library:
+
+template<typename T>
+C10_HOST_DEVICE inline c10::complex<T> log1p(const c10::complex<T> &z) {
```
Comment (anjali411) [inline c10/util/complex_math.h:218]: std::log is defined for complex so we could just call std::log(1+z)?

Comment (zasdfgbnm) [reply c10/util/complex_math.h:218]: std::log(1+z) has precision issue when z << 1:
https://stackoverflow.com/questions/49538185/what-is-the-purpose-of-numpy-log1p

Comment (anjali411) [reply c10/util/complex_math.h:218]: ah ok I see. can you add some perf numbers for log1p, existing log functions for complex

Comment (zasdfgbnm) [reply c10/util/complex_math.h:218]: Sure, will do it later today

Comment (zasdfgbnm) [reply c10/util/complex_math.h:218]:
```C++
#include <complex>
#include <iostream>
#include <chrono>

template<typename T>
inline std::complex<T> log1p_v1(const std::complex<T> &z) {
  // log1p(z) = log(1+z)
  // Let's define 1 + z = r*e^(i*a), then we have
  // log(r*e^(i*a)) = log(r) + i*a
  // but log(r) could have precision issue when |z| << 1, so we should really
  // be using log1p(r-1), where the r-1 should be computed in high precision.
  // to do so, we are doing the following transformation: (assuming z = x+iy)
  // r-1 = (r-1)*(r+1)/(r+1) = (r^2-1) / (r+1)
  //     = ((x+1)^2 + y^2 - 1) / (r+1)
  //     = (x^2 + y^2 + 2x) / (r+1)
  T x = z.real();
  T y = z.imag();
  std::complex<T> p1 = z + T(1);
  T r = std::abs(p1);
  T a = std::arg(p1);
  T rm1 = (x * x + y * y + x * T(2)) / (r + 1);
  return {std::log1p(rm1), a};
}

template<typename T>
inline std::complex<T> log1p_v2(const std::complex<T> &z) {
  return std::log(T(1) + z);
}

int main() {
  std::complex<float> input(0.5, 2.0);

  auto start = std::chrono::system_clock::now();
  for (int i = 0; i < 10000; i++) {
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
    log1p_v1(input);
  }
  auto end = std::chrono::system_clock::now();
  auto elapsed = end - start;
  std::cout << "time for v1: " << elapsed.count() << '\n';


  start = std::chrono::system_clock::now();
  for (int i = 0; i < 10000; i++) {
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
    log1p_v2(input);
  }
  end = std::chrono::system_clock::now();
  elapsed = end - start;
  std::cout << "time for v2: " << elapsed.count() << '\n';
}
```
v1 is 10% slower than v2:
```
time for v1: 22292445
time for v2: 20959112
```

File: c10/util/complex_math.h [hunk 2]
```diff
@@ -212,6 +212,28 @@ C10_HOST_DEVICE inline c10::complex<T> atanh(const c10::complex<T> &x) {
...
+  // to do so, we are doing the following transformation: (assuming z = x+iy)
+  // r-1 = (r-1)*(r+1)/(r+1) = (r^2-1) / (r+1)
+  //     = ((x+1)^2 + y^2 - 1) / (r+1)
+  //     = (x^2 + y^2 + 2x) / (r+1)
+  T x = z.real();
+  T y = z.imag();
+  c10::complex<T> p1 = z + T(1);
+  T r = std::abs(p1);
+  T a = std::arg(p1);
+  T rm1 = (x * x + y * y + x * T(2)) / (r + 1);
```
Comment (anjali411) [inline c10/util/complex_math.h:233]: `x**2 + y**2 = r**2` so I think you can write:
`T rm1 = (r*r + x * T(2)) / (r+T(1))`

File: c10/util/complex_math.h (hunk 1 above)
Comment (anjali411) [reply c10/util/complex_math.h:218]: hm that's a significant speed difference

File: c10/util/complex_math.h (hunk 2 above)
Comment (zasdfgbnm) [reply c10/util/complex_math.h]: Fixed, but the perf gain is little.

File: c10/util/complex_math.h (hunk 1 above)
Comment (zasdfgbnm) [reply c10/util/complex_math.h:218]: I tried `-Ofast` when compiling, the time is 90 vs 50, much significant. Maybe we can do:

```
if (abs(z) < 1) {
   v1
} else {
  v2
}
```
?

Comment (zasdfgbnm) [reply c10/util/complex_math.h:224]: I added a special case for using `std::log(T(1) + z)` when `z` is large enough.

File: c10/util/complex_math.h (hunk 2 above)
Comment (zasdfgbnm) [reply c10/util/complex_math.h:245]: Oh, no, `x` and `y` are real and imag for `z`, but `r` is the `r` for `z + 1`, so `x**2 + y**2 != r**2`

Comment (anjali411) [reply c10/util/complex_math.h:245]: ahh

**Comment by zasdfgbnm** [comment]
A good reference https://github.com/bminor/glibc/blob/43b1048ab9418e902aac8c834a7a9a88c501620a/sysdeps/ia64/fpu/s_log1pf.S

File: c10/util/complex_math.h (hunk 1 above)
Comment (anjali411) [reply c10/util/complex_math.h:224]: @zasdfgbnm we should probably add this for small z (prioritize accuracy over performance).

**Comment by Lezcano** [comment]
@zasdfgbnm what's the state of this PR? Could it be revived?
We were recently asked to make `logcumsumexp` support complex numbers, and to do that we would need complex support for `at::log1p`

**Pull request closed without merging**